controllers/userController.js

```diff
@@ -0,0 +1,180 @@
+  User.findOne({_id: req.params.id}, (err, result) => {
+    if (err) console.log(err);
+    res.render('profile', result);
+  });
+};
+
+
+// update user information to database, user
+// use mongoose save method instead of update method as mongoose middleware to hash password is only invoked on save method
+exports.change = (req, res) => {
+  // express validator to check profile form fields
+  req.checkBody('firstName', 'First name cannot be empty.').notEmpty();
+  req.checkBody('lastName', 'Last name cannot be empty.').notEmpty();
+  req.checkBody('email', 'Email cannot be empty.').notEmpty();
+  req.checkBody('username', 'Username cannot be empty.').notEmpty();
+  req.checkBody('password', 'Password cannot be empty.').notEmpty();
+  req.checkBody('passwordConfirm', 'Confirm your password again.').notEmpty();
+  req.checkBody('password', 'Passwords entered are not the same.').equals(req.body.passwordConfirm);
+
+  let errors = req.validationErrors();
+  if (errors) {
+    res.render('profile', {'errors': errors});
+  } else {
+    User.findOne({_id: req.params.id}, (err, result) => {
+      // console.log(result);
+      if (err) console.log(err);
+
+      result.firstName = req.body.firstName;
+      result.lastName = req.body.lastName;
+      result.email = req.body.email;
+      result.username = req.body.username;
+      result.password = req.body.password;
+      // validate password and cannot be empty field
+
+      result.save((err, updatedResult) => {
+        if (err) console.log(err);
+        // console.log(updatedResult);
+        req.flash('success', 'Credentials saved.');
+        res.redirect('/profile/' + req.params.id);
+      });
+    });
+  }
+};
+
+
+// delete search and analyzed data from database, searchList and analyzedList, through ajax request
+exports.deleteSearch = (req, res) => {
+  let query = {_id: req.params.id};
+  // console.log(query);
+  SearchList.remove(query, (err) => {
+    if (err) console.log(err);
+    console.log("deleted from searchList");
+    res.sendStatus(200);
+  });
+};
+
+exports.deleteAnalyzed = (req, res) => {
+  let query = {_id: req.params.id};
+  // console.log(query);
+  AnalyzedList.remove(query, (err) => {
+    if (err) console.log(err);
+    console.log("deleted from analyzedList");
+    res.sendStatus(200);
+  });
+};
+
+exports.deleteSaved = (req, res) => {
+  let query = {_id: req.params.id};
+  // console.log(query);
+  SavedList.remove(query, (err) => {
+    if (err) console.log(err);
+    console.log("deleted from savedList");
+    res.sendStatus(200);
+  });
+};
+
+
+// create, save combinations which user chooses
+exports.saveAnalyzed = (req, res) => {
+  // console.log(req.params.id);
+  // console.log(req.body);
+  SavedList.findOne({username: req.params.id, item: req.body.item}, (err, data) => {
+    if (err) console.log(err);
+    if (data) {
+      // add to database if entry is already in database
+
+      // check to see if combination is already present
```

Comment:
you can skip this step by putting the logic check in `SavedList` model instead. https://stackoverflow.com/questions/16882938/how-to-check-if-that-data-already-exist-in-the-database-during-update-mongoose